Redesign buttons in settings tables.

shuklamaneesh23 opened the pull request:
<!-- Describe your pull request here.-->

This PR redesigns various button present in settings.
Please see the screenshots attached which describe this PR more better.

Fixes: #34200.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

### Pre-commit change to update the existing delete svg:

| Before | After |
|------|------|
| <img width="156" alt="Screenshot 2025-04-02 at 5 59 50 PM" src="https://github.com/user-attachments/assets/d8f7334d-790f-4d85-b5fe-d7f1bf0edd32" /> | <img width="161" alt="Screenshot 2025-04-02 at 5 52 50 PM" src="https://github.com/user-attachments/assets/7ca3e884-2104-4837-9b9a-8c6bb1dc7fa6" /> | 
| <img width="133" alt="Screenshot 2025-04-02 at 5 59 55 PM" src="https://github.com/user-attachments/assets/a0601d2d-d8d2-4d94-a0a1-b619040e6f9e" /> | <img width="188" alt="Screenshot 2025-04-02 at 5 52 59 PM" src="https://github.com/user-attachments/assets/f0bf0810-d058-4c40-a0ba-f1db6d04854f" /> | 
| <img width="172" alt="Screenshot 2025-04-02 at 6 00 03 PM" src="https://github.com/user-attachments/assets/25290fc6-b2df-4e68-afce-5c709578da66" /> | <img width="184" alt="Screenshot 2025-04-02 at 5 53 10 PM" src="https://github.com/user-attachments/assets/a982203f-e264-4485-9ae5-84cf4d26b8f5" /> | 


**Screenshots and screen captures:**

### Main changes:

| **Location**                  | **Before (At 16px)** | **After (At 12px)** | **After (At 16px)** | **After (At 20px)** |
|--------------------------------|----------------------|----------------------|----------------------|----------------------|
| Custom emoji                  | ![Before](https://github.com/user-attachments/assets/b5493c64-3e18-462e-9582-f60828d5b93f) | ![12px](https://github.com/user-attachments/assets/db61da4e-ea7f-4efe-bf87-98a1bea45af5) | ![16px](https://github.com/user-attachments/assets/0c1d5d36-77fd-48c7-a1e4-ab28e2aa5060) | ![20px](https://github.com/user-attachments/assets/520a4f1c-93f8-413f-a926-f38dc1b032a1) |
| Linkifier                     | ![Before](https://github.com/user-attachments/assets/f76d1ed4-5426-4803-8edc-bb0886d87865) | ![12px](https://github.com/user-attachments/assets/c7d62c59-22dc-47d2-99d3-e5195eb58dd8) | ![16px](https://github.com/user-attachments/assets/bacdd03c-6b7f-40e5-b35a-03dc12c5e5cb) | ![20px](https://github.com/user-attachments/assets/27eb6795-7d8b-4a96-971b-245c8124b46b) |
| Code playground               | ![Before](https://github.com/user-attachments/assets/1280b1df-43fe-4b8e-8af0-35e35aac3536) | ![12px](https://github.com/user-attachments/assets/190c8ba2-d313-4008-b0a8-63eefcec0b6d) | ![16px](https://github.com/user-attachments/assets/5aabb945-f5ab-4afa-b48c-5a1e38033877) | ![20px](https://github.com/user-attachments/assets/0cd8e9f2-bcc3-4acd-a631-812aff886f4e) |
| Manage & Deactivate User       | ![Before](https://github.com/user-attachments/assets/a2483e64-a5a5-4058-9351-79882dcde28f) | ![12px](https://github.com/user-attachments/assets/4af6f057-1d11-4158-aa37-0fdd7adc0010) | ![16px](https://github.com/user-attachments/assets/24d79eb1-d3de-4e93-88d1-021dc55beaa8) | ![20px](https://github.com/user-attachments/assets/ba06800e-c07b-429a-9531-255802c53ad1) |
| Reactivate User               | ![Before](https://github.com/user-attachments/assets/43219884-51c4-4656-bc6b-da215d9505ac) | ![12px](https://github.com/user-attachments/assets/aa255851-afe9-422d-a0b3-ea8a31f5c0de) | ![16px](https://github.com/user-attachments/assets/ef90a7cf-9573-41f6-ac14-a9911edf8f86) | ![20px](https://github.com/user-attachments/assets/3a8c7c36-945d-4b23-9ed5-50cf8b80bb7e) |
| Invite User to Org            | ![Before](https://github.com/user-attachments/assets/a62acd33-d270-477d-8f94-222317212a58) | ![12px](https://github.com/user-attachments/assets/2415a46c-3b7f-49b0-b3c7-9bb46a55287c) | ![16px](https://github.com/user-attachments/assets/6b5cd898-4171-4dce-b836-c3800b9a6f0a) | ![20px](https://github.com/user-attachments/assets/7015c40f-d961-474d-baf3-4b7f7628f823) |
| Custom Profile                | ![Before](https://github.com/user-attachments/assets/d4485412-39aa-4d0c-8fa3-9c0b2d5100d5) | ![12px](https://github.com/user-attachments/assets/1dbe1799-bbcf-4e76-8169-b76db5b617d9) | ![16px](https://github.com/user-attachments/assets/b5cb189c-08d8-4a64-b554-0b30bfbccc51) | ![20px](https://github.com/user-attachments/assets/4681d6d3-dc9d-4fc9-8ae3-f0a5ad70e614) |
| Default Channel               | ![Before](https://github.com/user-attachments/assets/76119a02-7bb6-4000-a92e-83bf5b6f6b3a) | ![12px](https://github.com/user-attachments/assets/fd585cfd-8ef0-4dd6-8b5f-876f07a5a782) | ![16px](https://github.com/user-attachments/assets/7fcc40db-05bc-4465-9278-09fa6b6a136f) | ![20px](https://github.com/user-attachments/assets/37833d61-af01-4029-8e0e-5b85c24f9bba) |
| Data Export Delete            | ![Before](https://github.com/user-attachments/assets/6ea9cca9-960d-42ef-8420-e2c6b1c287df) | ![12px](https://github.com/user-attachments/assets/19234b96-e883-42ce-be98-654f8747f383) | ![16px](https://github.com/user-attachments/assets/06506943-ab2f-49e2-8d13-f919029ffd49) | ![20px](https://github.com/user-attachments/assets/19bf9d15-3292-47ba-b220-de52c54715aa) |
| Resend and Revoke | <img width="206" alt="Screenshot 2025-03-04 at 9 47 33 PM" src="https://github.com/user-attachments/assets/4f96777f-41cd-42a8-a575-98606ab59bf1" /> | <img width="70" alt="Screenshot 2025-04-02 at 10 40 36 PM" src="https://github.com/user-attachments/assets/d3148ec5-33e8-4955-8220-74f1e2640726" /> | <img width="86" alt="Screenshot 2025-04-02 at 10 41 35 PM" src="https://github.com/user-attachments/assets/20930165-553a-4ec1-aecd-f32c345eac33" /> | <img width="106" alt="Screenshot 2025-04-02 at 10 41 59 PM" src="https://github.com/user-attachments/assets/c862f495-7589-46ca-ac9b-a9072e694d16" /> |
| Sent and Revoke | <img width="156" alt="Screenshot 2025-03-06 at 9 30 29 PM" src="https://github.com/user-attachments/assets/79a329eb-8c38-454a-838f-19ed75938d3f" /> | <img width="80" alt="Screenshot 2025-04-02 at 10 40 50 PM" src="https://github.com/user-attachments/assets/3f24140c-a815-4228-911b-801f0160866b" /> | <img width="95" alt="Screenshot 2025-04-02 at 10 41 41 PM" src="https://github.com/user-attachments/assets/2b6c6b75-65f3-40d0-8b4d-9cc74c41fcc8" /> | <img width="112" alt="Screenshot 2025-04-02 at 10 42 05 PM" src="https://github.com/user-attachments/assets/0db12a17-a6a9-4ae5-9f73-6103ea66aef7" /> |




<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x]  Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks! I don't think the trash can looks quite right. Can you use the SVG I added to the issue description?

alya commented:
I added some notes to the issue about the "Resend" / "Sent!" buttons. I had only made a link invitation, so I didn't see them when I was writing up the issue.

alya commented:
Oh, and let's put the "Revoke" button to the left of "Resend", so that they line up in a column.

![Screenshot 2025-03-31 at 23 14 16@2x](https://github.com/user-attachments/assets/084fb372-6f93-47cf-ad73-d384833a549a)

alya commented:
Looks good other than those points in manual testing.

alya commented:
Replace -- in your screenshots, it looks like something is wrong with the second line inside. They should look identical: http://lucide.dev/icons/trash-2

alya commented:
Let's still reorder the icon buttons, so that they line up between email invitations and link invitations.

shuklamaneesh23 commented:
> Let's still reorder the icon buttons, so that they line up between email invitations and link invitations.

Done @alya , please take a look now and let me know if i can improve this further.
Thanks!

alya commented:
The deactivate / reactivate user button needs to update properly when a user is deactivated / reactivated. Currently, the icon changes correctly, but the color and the tooltip do not.

alya commented:
For custom emoji, let's change the tooltip "Delete" -> "Deactivate" to better match the confirmation modal / what actually happens. This might require a help center update.

alya commented:
For system bots, let's change the tooltip "This bot cannot be edited." -> "This bot cannot be managed." Looks like we forgot to do that earlier.

alya commented:
Looks good to me in manual testing otherwise!

shuklamaneesh23 commented:
Hi @alya , i've reworked this PR. Please play with it again.

Updates:

- Made the deactivate / reactivate user button update properly when a user is deactivated / reactivated.
- For system bots, changed the tooltip "This bot cannot be edited." -> "This bot cannot be managed."
- For custom emoji, changed the tooltip "Delete" -> "Deactivate" 
- Added explanatory tooltips for "Resend" and "Revoke".
- Made checkmark confirmation not clickable, now it's a regular pointer cursor and no hover/click animation, but the icon would not look disabled. 

Also, i checked the `help-center` in order to update that as you said but i think the current state is fine, please have a look at them and let me know if they still needs to be updated:

<img width="774" alt="Screenshot 2025-04-03 at 7 49 57 PM" src="https://github.com/user-attachments/assets/b8262b92-13e9-4f24-a68d-4969696e690d" />
<img width="761" alt="Screenshot 2025-04-03 at 7 54 17 PM" src="https://github.com/user-attachments/assets/9423c9af-a866-46d0-be1e-fb14b8688f29" />

alya commented:
Since we're replacing the labeled buttons with icons, we need to use the same way to describe them as other icon buttons.

alya commented:
And the label in the help center should always match the tooltip, so "trash" is not correct.

alya commented:
Is that using the old trash can icon? The icon needs to be updated to the new one everywhere that we changed the UI.

alya commented:
@sahil839 Could you please review this one? Should be ready for you other than the help center changes. I haven't re-tested the latest version.

sahil839 commented on web/templates/components/icon_button.hbs:
```diff
@@ -1,4 +1,4 @@
-<button type="button" {{#if id}}id="{{id}}"{{/if}} class="{{#if custom_classes}}{{custom_classes}} {{/if}}icon-button {{#if squared}}icon-button-square {{/if}}icon-button-{{intent}} {{#if hidden}}hide{{/if}}" {{#if data-tooltip-template-id}}data-tooltip-template-id="{{data-tooltip-template-id}}"{{/if}} tabindex="0" {{#if aria-label}}aria-label="{{aria-label}}"{{/if}}
+<button type="{{#if type}}{{type}}{{else}}button{{/if}}" {{#if id}}id="{{id}}"{{/if}} class="{{#if custom_classes}}{{custom_classes}} {{/if}}icon-button {{#if squared}}icon-button-square {{/if}}icon-button-{{intent}} {{#if hidden}}hide{{/if}}" {{#if data-tooltip-template-id}}data-tooltip-template-id="{{data-tooltip-template-id}}"{{/if}} tabindex="0" {{#if aria-label}}aria-label="{{aria-label}}"{{ ...
```
I guess in one of the PRs for redesigning the buttons, a separate handler was added for the button as it used form submission. But here we already have the click handler, so I guess we can skip this change.

sahil839 commented on web/templates/settings/uploaded_files_list.hbs:
```diff
@@ -20,16 +20,22 @@
-            <a type="submit" href="/user_uploads/{{path_id}}" class="button rounded small sea-green tippy-zulip-delayed-tooltip download-attachment" data-tippy-content="{{t 'Download' }}" download>
-                <i class="fa fa-download sea-green" aria-hidden="true"></i>
-            </a>
+            <a type="submit" href="/user_uploads/{{path_id}}" id="hidden-attachment-download" download></a>
```
Is there a way to directly let the click handler for button handle the download instead of having this `a` element?

shuklamaneesh23 replied:
No, the `<button>` tag doesn't support the `download` attribute. There has been a discussion on the [CZO](https://chat.zulip.org/#narrow/channel/6-frontend/topic/Migration.20of.20sea-green.20to.20action-button-quiet-brand.2E/near/2115891) regarding this change that how we want to proceed here.

sahil839 commented on web/templates/settings/invites_list_admin.hbs:
```diff
@@ -3,11 +3,12 @@
+      attention="quiet"
+      intent="brand"
+      custom_classes="user-settings-invite-user-label invite-user-link"
+      }}
```
I think this change can be in the next commit which makes other changes in "Invitations" panel.

shuklamaneesh23 replied:
> This button is not aligned perfectly with the tabs above at the left. Might be a small CSS change.

This button was algned with the `<h3>Invitations</h3>` previously. So, i'm keeping that alignment with that heading element rather than going with the tab above since both are in different divs.

| Before | After |
|------|-------|
| <img width="547" alt="Screenshot 2025-04-04 at 1 45 40 PM" src="https://github.com/user-attachments/assets/ea860afd-2573-4f03-bf64-d7f11849850c" /> | <img width="470" alt="Screenshot 2025-04-04 at 1 38 05 PM" src="https://github.com/user-attachments/assets/d441b2ee-222d-4b24-9dc0-a6f0ba56da23" /> |

sahil839 replied:
Then we can align tabs with the heading, button, etc.

sahil839 commented on web/styles/app_components.css:
```diff
@@ -812,6 +812,16 @@ input.settings_text_input {
+            vertical-align: middle;
+        }
+
+        .check:disabled {
```
This change is applied to all tables, so we should make sure it does not affect other tables.

shuklamaneesh23 replied:
The `check` class is only available in the Invitations table and hence it won't affect any of the others table.

sahil839 replied:
Might still be better to make this CSS more specific to invitations table.

shuklamaneesh23 commented on web/styles/app_components.css:
```diff
@@ -812,6 +812,16 @@ input.settings_text_input {
+        /* The vertical-align was originally set to 'top', causing misalignment in this section.
+        Changing it to 'middle' aligns the icon properly. */
+        .zulip-icon::before {
+            vertical-align: middle;
```
I can't determine the reason for this mis-alignment. If i'm putting the same icon-button in any other action columns, it is perfectly aligned.

However, while doing `hit and trial` with different css properties in the inspect mode, i found that there are two properties changing which fixing the alignment and those are:
- `vertical-align:top` to `vertical-align: middle`
- `font-family: zulip-icons` to `font-family: sans-serif`

So, i went with first one and that was looking good in my system.

<details>
<summary> See this </summary>
<img width="300" alt="Screenshot 2025-04-04 at 4 56 19 PM" src="https://github.com/user-attachments/assets/3ec9a922-0ec3-4484-9ff6-407e341812c5" />

</details>

However, i'm now considering to change the `font-family` as you reported that the previous was not working for you, please have a look now:

<img width="657" alt="Screenshot 2025-04-04 at 6 01 19 PM" src="https://github.com/user-attachments/assets/09cc4fb2-ee68-4971-be74-119922ed7f10" />

sahil839 replied:
I think there can be a better solution than changing font-family. I guess this is happening due to `inline-block` property on button which does not align the icon centrally in the button. We can probably instead use `inline-flex` property like we do for buttons in "Users" list.

And I also noticed we need this change for all the tables where `inline-block` is added to buttons.

shuklamaneesh23 commented:
Hi @sahil839 , thanks for your feedback.
I've ,made the code changes wherever it applies and replied to your comments wherever it required.
Please have another look.

sayamsamal commented:
The language in the "Manage pending invitations" seems off to me.

alya commented on help/custom-profile-fields.md:
```diff
@@ -38,7 +38,7 @@ methods][authentication-production] documentation for details.
 {settings_tab|profile-field-settings}
 
-1. In the **Actions** column, click the **pencil** (<i class="fa fa-pencil"></i>)
+1. In the **Actions** column, click the **pencil** (<i class="zulip-icon zulip-icon-edit"></i>)
```
same below

alya commented on help/manage-your-uploaded-files.md:
```diff
@@ -42,7 +42,7 @@ You can sort your uploaded files by name, upload date, message ID, and size.
 {settings_tab|uploaded-files}
 
 1. In the **Actions** column, click the **download file**
```
```suggestion
1. In the **Actions** column, click the **download**
```

alya commented on help/manage-your-uploaded-files.md:
```diff
@@ -53,7 +53,7 @@ You can sort your uploaded files by name, upload date, message ID, and size.
 {settings_tab|uploaded-files}
 
 1. In the **Actions** column, click the **delete file**
```
```suggestion
1. In the **Actions** column, click the **delete**
```

alya commented:
> The language in the "Manage pending invitations" seems off to me.

Yeah, that's not really a sentence. We just need to make separate instruction blocks. So, under "Manage pending invitations", make subsections for "Revoke an invitation" and "Resend an invitation". Put specific instructions for each in the usual style.

shuklamaneesh23 commented:
How does this look @alya ?

<details>
<summary>Screenshot </summary>
<img width="774" alt="Screenshot 2025-04-07 at 2 25 25 AM" src="https://github.com/user-attachments/assets/d74f31cf-4649-4be9-b8b0-a0aa49bbc978" />
</details>

alya commented on help/invite-new-users.md:
```diff
@@ -94,17 +94,42 @@ Organization owners can revoke or resend any invitation or reusable
 
+### Revoke an invitation
+
+To revoke a pending invitation:
+
```
```suggestion
```

alya commented on help/invite-new-users.md:
```diff
@@ -94,17 +94,42 @@ Organization owners can revoke or resend any invitation or reusable
 
+### Resend an invitation
+
+To resend a pending invitation:
+
```
```suggestion
```

alya replied:
```suggestion
```

alya commented:
Getting close!

sahil839 commented on web/templates/settings/admin_profile_field_list.hbs:
```diff
@@ -34,13 +34,20 @@
-        <button class="button rounded small delete button-danger tippy-zulip-delayed-tooltip" data-tippy-content="{{t 'Delete' }}" data-profile-field-id="{{id}}">
-            <i class="fa fa-trash-o" aria-hidden="true"></i>
-        </button>
+    <td class="actions" data-profile-field-id="{{id}}">
```
This should be moved to `tr` for all the tables.

shuklamaneesh23 commented:
Hi @alya  and @sahil839 , i've updated the `help-center` docs as well as the code.
Please have a look.
Thanks!

sahil839 commented on web/styles/app_components.css:
```diff
@@ -812,6 +812,10 @@ input.settings_text_input {
         white-space: nowrap;
+
+        button {
+            display: inline-flex;
```
We should add `min-height` here like we have for users table, so that row has some minimum height.

sahil839 replied:
Is this style needed only for icon buttons? I guess so.

shuklamaneesh23 replied:
Yes, i have made the selector more specific.

sahil839 commented on web/styles/settings.css:
```diff
@@ -1341,26 +1341,6 @@ label.preferences-radio-choice-label {
     align-items: center;
 }
 
-.invite-user-link {
```
This change should be in commit which changes the button.

sahil839 commented on web/styles/buttons.css:
```diff
@@ -423,6 +423,10 @@
     }
+
+    &.hide {
+        display: none !important;
```
Or you can call `hide()` and `show()` function from jquery as well.

sahil839 commented on web/src/attachments_ui.ts:
```diff
@@ -196,6 +196,10 @@ export function set_up_attachments(): void {
     });
 
+    $("#uploaded_files_table").on("click", ".download-attachment", () => {
+        $("#hidden-attachment-download")[0]?.click();
```
For some reason, wrong file is being downloaded. I uploaded three files, and same file is being downloaded while clicking on download button of all three files.

sayamsamal replied:
Mostly to do with the `#hidden-attachment-download` id being repeated for every attachment, which we shouldn't be having. A better method would be to use `this` and track the closest element with the attachment.

sahil839 commented on web/src/settings_invites.ts:
```diff
@@ -225,8 +225,22 @@ function do_resend_invite({$row, invite_id}: {$row: JQuery; invite_id: string}):
+            $resend_button
+                .removeClass("resend icon-button-neutral")
+                .addClass("check icon-button-success");
+            $resend_button.find("i").removeClass("zulip-icon-send-dm").addClass("zulip-icon-check");
```
Any reason to not have both icons present and hide and show it accordingly? Seems easy to understand and will help in avoiding future bugs.

sahil839 reviewed (commented):
Posted one comment. Looks good otherwise.

shuklamaneesh23 commented:
Hi @sahil839 , could you please take a look now?
Thanks!

sahil839 commented:
I missed one thing before, order of resend and revoke buttons is changed in Invitations panel. Was this discussed? Looks good otherwise.

sahil839 commented:
Ok. Looks good then. I have not reviewed the docs changes.
